apps/condo/domains/miniapp/schema/preprocessors/B2BAppAccess.js

```diff
@@ -0,0 +1,470 @@
+ * @return {*|DocumentNode}
+ *
+ * @example
+ * Input: ['ticket', 'organization', 'id']
+ * Output: gql`
+ *         query getAllTickets ($where: TicketWhereInput, $first: Int = 1) {
+ *             objs: allTickets(where: $where, first: $first) { organization { id } }
+ *         }
+ *     `
+ */
+const generateGql = (pathToOrganizationId) => {
+    if (!pathToOrganizationId || !isArray(pathToOrganizationId) || isEmpty(pathToOrganizationId)) throw new Error('"pathToOrganizationId" should not be empty array')
+    if (pathToOrganizationId.length < 3) throw new Error(`To generate gql "pathToOrganizationId" must contain at least three elements! But was ${pathToOrganizationId}`)
+
+    const gqlKey = pathToOrganizationId.join(':')
+
+    if (GENERATED_SCHEMAS.has(gqlKey)) {
+        return GENERATED_SCHEMAS.get(gqlKey)
+    }
+
+    const listKey = pathToOrganizationId[0].charAt(0).toUpperCase() + pathToOrganizationId[0].slice(1)
+
+    const generatedGql = gql`
+        query getAll${pluralize.plural(listKey)} ($where: ${listKey}WhereInput, $first: Int = 1) {
+            objs: all${pluralize.plural(listKey)}(where: $where, first: $first) ${generateGqlDataPart(pathToOrganizationId.slice(1))}
+        }
+    `
+    GENERATED_SCHEMAS.set(gqlKey, generatedGql)
+
+    return generatedGql
+}
+
+const canManageByServiceUser = async ({ authentication: { item: user }, listKey, originalInput, itemId, operation, context }, schemaConfig) => {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (!listKey) return false
+
+    const pathToOrganizationId = get(schemaConfig, 'pathToOrganizationId', ['organization', 'id'])
+
+    let organizationId
+
+    if (operation === 'create') {
+        if (pathToOrganizationId.length === 1) {
+            organizationId = get(originalInput, [pathToOrganizationId[0]], null)
+        } else if (pathToOrganizationId.length === 2) {
+            organizationId = get(originalInput, [pathToOrganizationId[0], 'connect', pathToOrganizationId[1]], null)
+        } else if (pathToOrganizationId.length === 3) {
+            const parentObjectId = get(originalInput, [pathToOrganizationId[0], 'connect', 'id'], null)
+            if (!parentObjectId) return false
+
+            const [parentObject] = await execGqlWithoutAccess(context, {
+                query: generateGql(pathToOrganizationId),
+                variables: {
+                    where: { id: parentObjectId },
+                    first: 1,
+                },
+                dataPath: 'objs',
+            })
+
+            if (!parentObject) return false
+
+            organizationId = get(parentObject, [pathToOrganizationId[1], pathToOrganizationId[2]])
+        } else {
+            throw new Error(`You should implement "canManageByServiceUser" when "pathToOrganizationId" have depth ${pathToOrganizationId.length}`)
+        }
+    } else if (operation === 'update') {
+        const item = await getById(listKey, itemId)
```

Comment:
This can also be simplified:
1. If length is one, take itemId directly.
2. If length is more, build a query and unpack it

Reply:
Please also note (probably add test for it?), that `itemId` is not always defined. In case of bulk-requests (`updateProperties` for example) keystone uses array named `itemIds` instead (for reference check out receipts in billing domain)

It'll be good to support bulk requests, since some miniapps update information by CRON, and we can save it from some request-spam

Reply:
It's not that hard to implement as well:
1. In case of update `itemId` becomes `[itemId]` (bulk of one). Then you can generate same query, but use `id_in` instead

Reply:
I also concerned about relations, since it's not checked there. Should we protect relations from being changed?

Since right now it's possible to change it from miniapp (change organization field on contact and other things like that)

Reply:
> This can also be simplified:
> 
> 1. If length is one, take itemId directly.
> 2. If length is more, build a query and unpack it

done

Reply:
> Please also note (probably add test for it?), that `itemId` is not always defined. In case of bulk-requests (`updateProperties` for example) keystone uses array named `itemIds` instead (for reference check out receipts in billing domain)
> 
> It'll be good to support bulk requests, since some miniapps update information by CRON, and we can save it from some request-spam

Please, not now)